jsonschema-core.md

```diff
@@ -1386,13 +1375,15 @@ applicator keywords or with location-reserving keywords such as
 be `$defs` and the standard applicators from this document or
 implementation-specific custom keywords.
 
-Multi-level structures of unknown keywords are capable of introducing nested
-subschemas, which would be subject to the processing rules for `$id`. Therefore,
-having a reference target in such an unrecognized structure cannot be reliably
-implemented, and the resulting behavior is undefined. Similarly, a reference
-target under a known keyword, for which the value is known not to be a schema,
-results in undefined behavior in order to avoid burdening implementations with
-the need to detect such targets.[^10]
+Note that single-level custom keywords with identical syntax and semantics to
+`$defs` do not allow for any intervening `$id` keywords, and therefore will
+behave correctly under implementations that attempt to use any reference target
+as a schema. However, this behavior is implementation-specific and MUST NOT be
+relied upon for interoperability.
```

Comment:
I find this paragraph confusing. The word "intervening" doesn't make sense in this context. I also found it awkward that it starts by saying it will behave correctly, then says it must not be relied upon. I think it would be more clear if it was the other way around.

I think this was originally about unknown keywords. Since unknown keywords aren't allowed anymore, it becomes about custom vocabulary defined keywords. Therefore, I think this is saying that you can't use `$id` in a schema in a custom keyword. I don't think that's what we want. `$id` should work in any known schema, even in custom keywords.

Reply:
This paragraph was just moved up.  It was previously at line 1404 (which you can see just below).

> I also found it awkward that it starts by saying it will behave correctly, then says it must not be relied upon.

The text is "will behave correctly under implementations that attempt to use any reference target as a schema" which is correct.

> I think this was originally about unknown keywords.

It's not just about unknown keywords.  It's also about keywords that can define structure but not subschemas, like `default` or `const` (or custom keywords with structured data).  The point of this paragraph is to say that `$id` in any place where a schema is not expected is to treated as data, not an identifier.

Reply:
Yes, I noticed that it was moved. I'm not arguing about its correctness. I can't even say anything about its correctness at this point because I don't understand what it's trying to say.

> It's not just about unknown keywords. It's also about keywords that can define structure but not subschemas, like `default` or `const` (or custom keywords with structured data).

I don't see how that paragraph can be understood to be about that. It appears this only applies to custom keywords with identical semantics to `$defs`.

> The point of this paragraph is to say that `$id` in any place where a schema is not expected is to treated as data, not an identifier.

That's well stated and clear. I suggest replacing the paragraph with some variation of that sentence.

Reply:
Yeah, reading it again, I don't think it's necessary.  It looks like it's just trying to explain that you _can't_ have an `$id` between the keyword simply because of the structure of the keyword.  I'm not sure that needs explaining, so I'll just remove it.

I'm not sure if we need to keep the "requirement" about "MUST NOT be relied upon". It's definitely a requirement on schema authors, but it's a weird one at that.

The next paragraph actually covers exactly what I was saying:

> A reference target under a keyword for which the value is known not to be a schema results in undefined behavior in order to avoid burdening implementations with the need to detect such targets.